clients/venice-thin-client/src/main/java/com/linkedin/venice/client/schema/RouterBackedSchemaReader.java

```diff
@@ -213,38 +223,40 @@ private void refreshAllValueSchema() throws VeniceClientException {
         LOGGER.info("Got null for request path: {}", requestPath);
         return;
       }
+
       MultiSchemaResponse schemaResponse = OBJECT_MAPPER.readValue(response, MultiSchemaResponse.class);
-      if (!schemaResponse.isError()) {
-        // Update local cache
-        int latestSchemaId = SchemaData.INVALID_VALUE_SCHEMA_ID;
-        MultiSchemaResponse.Schema[] entries = schemaResponse.getSchemas();
-        for (MultiSchemaResponse.Schema schema: entries) {
-          if (SchemaData.INVALID_VALUE_SCHEMA_ID == latestSchemaId || latestSchemaId < schema.getId()) {
-            latestSchemaId = schema.getId();
-          }
-          final String schemaStr = schema.getSchemaStr();
-          Schema writerSchema = preemptiveSchemaVerification(
-              // Use the "LOOSE" mode here since we might have registered schemas that do not pass the STRICT validation
-              // and that is allowed for now
-              parseSchemaFromJSONLooseValidation(schemaStr),
-              schemaStr,
-              schema.getId());
-          valueSchemaMap.put(schema.getId(), writerSchema);
-          valueSchemaMapR.put(writerSchema, schema.getId());
+      if (schemaResponse.isError()) {
+        throw new VeniceClientException(
+            "Received an error while fetching schema. " + getExceptionDetails(requestPath) + ", error message: "
+                + schemaResponse.getError());
+      }
+
+      int latestSchemaId = SchemaData.INVALID_VALUE_SCHEMA_ID;
+      for (MultiSchemaResponse.Schema schema: schemaResponse.getSchemas()) {
+        if ((latestSchemaId == SchemaData.INVALID_VALUE_SCHEMA_ID || latestSchemaId < schema.getId())
+            && schema.getId() != schemaResponse.getSuperSetSchemaId()) {
+          latestSchemaId = schema.getId();
         }
-        if (schemaResponse.getSuperSetSchemaId() != -1) {
+        String schemaStr = schema.getSchemaStr();
+        Schema writerSchema = preemptiveSchemaVerification(
+            // Use the "LOOSE" mode here since we might have registered schemas that do not pass the STRICT validation
+            // and that is allowed for now
+            parseSchemaFromJSONLooseValidation(schemaStr),
+            schemaStr,
+            schema.getId());
+        valueSchemaMap.put(schema.getId(), writerSchema);
+        valueSchemaMapR.put(writerSchema, schema.getId());
+      }
+
+      synchronized (this) {
+        if (supersetSchemaEnabled && schemaResponse.getSuperSetSchemaId() != SchemaData.INVALID_VALUE_SCHEMA_ID) {
           latestSchemaId = schemaResponse.getSuperSetSchemaId();
           latestValueSchemaEntry.set(new SchemaEntry(latestSchemaId, valueSchemaMap.get(latestSchemaId)));
-        } else {
-          if (SchemaData.INVALID_VALUE_SCHEMA_ID != latestSchemaId
-              && (latestValueSchemaEntry.get() == null || latestValueSchemaEntry.get().getId() < latestSchemaId)) {
+        } else if (latestSchemaId != SchemaData.INVALID_VALUE_SCHEMA_ID) {
```

Comment:
I'm guessing this `filter_schema_without_property` config will be a store config that's updated manually via admin tool by Venice devs/oncall upon user request?

Reply:
I was thinking it will be a client config, which is configured by Frame when constructing a client factory.

Reply:
Added configurable schema filter, so that it can extended later on. With this change `RouterBackedSchemaReader` searches for the largest id preferred schema first, and if not found resorts to the old strategy.